src/Algebra/Linear/Properties/VectorSpace.agda

```diff
@@ -0,0 +1,191 @@
+                 foldr (_+_ ∘ (v ∙_) ∘ g) (v ∙ x₀) xs
+foldr-homo-∙ _ [] = ∙-congˡ (≈ᴹ-reflexive Eq.refl)
+foldr-homo-∙ {v} {x₀} {g} g-cong (x ∷ xs) = begin⟨ setoid ⟩
+  v ∙ (g x +ᴹ foldr (_+ᴹ_ ∘ g) x₀ xs)        ≈⟨ ∙-distrib-+ ⟩
+  v ∙ g x + v ∙ foldr (_+ᴹ_ ∘ g) x₀ xs       ≈⟨ +-congˡ (foldr-homo-∙ g-cong xs) ⟩
+  foldr (_+_ ∘ (v ∙_) ∘ g) (v ∙ x₀) (x ∷ xs) ∎
+
+u∙-homo : ∀ {u} → IsModuleHomomorphism (u ∙_)
+u∙-homo = record
+  { isBimoduleHomomorphism = record
+      { +ᴹ-isGroupHomomorphism = record
+          { isMonoidHomomorphism = record
+              { isMagmaHomomorphism = record
+                  { isRelHomomorphism = record
+                      { cong = ∙-congˡ
+                      }
+                  ; homo = λ x y → ∙-distrib-+
+                  }
+              ; ε-homo = ∙-idʳ
+              }
+          ; ⁻¹-homo = λ x → ∙-homo-⁻¹
+          }
+      ; *ₗ-homo = λ r x → ∙-comm-*ₗ
+      ; *ᵣ-homo = λ r x → ∙-comm-*ᵣ
+      }
+  }
+
+------------------------------------------------------------------------
+-- Properties of linear maps from vectors to their underlying scalars.
+module _ (lm : LinearMap mod ⟨module⟩) where
+
+  open LinearMap lm
+  open ModHomo   mod ⟨module⟩ (LinearMap.homo lm)
+
+  vred : (V → S) → List V → S
+  vred g = foldr (_+_ ∘ uncurry _*_ ∘ < g , f >) 0#
+
+  foldr-homo : (g : V → S) → (xs : List V) → f (vgen g xs) ≈ vred g xs
+  foldr-homo g []       = 0ᴹ-homo
+  foldr-homo g (x ∷ xs) = begin⟨ setoid ⟩
+    f (h x (foldr h 0ᴹ xs))
+      ≈⟨ +ᴹ-homo (g x *ₗ x) (foldr h 0ᴹ xs) ⟩
+    f (g x *ₗ x) + f (foldr h 0ᴹ xs)
+      ≈⟨ +-congʳ (*ₗ-homo (g x) x) ⟩
+    g x * f x + f (foldr h 0ᴹ xs)
+      ≈⟨ +-congˡ (foldr-homo g xs) ⟩
+    g x * f x + vred g xs
+      ∎
+    where
+    h = _+ᴹ_ ∘ vscale g
+
+  vSum : List V → V
+  vSum xs = foldr _+ᴹ_ 0ᴹ xs
+
+  fScale : V → V
+  fScale = vscale f
+
+  fGen : List V → V
+  fGen = vgen f
+
+  f≈v∙ : ∀ {a} → f a ≈ lmToVec lm ∙ a
+  f≈v∙ {a} = sym (begin⟨ setoid ⟩
+    v′ ∙ a ≈⟨ ∙-comm ⟩
+    a ∙ v′ ≈⟨ foldr-homo-∙ (vscale-cong f ⟦⟧-cong) basisSet ⟩
+    foldr (_+_ ∘ (a ∙_) ∘ fScale) (a ∙ 0ᴹ) basisSet
+      ≈⟨ foldr-cong (λ {y≈z _ → +-congˡ y≈z}) ∙-idʳ basisSet ⟩
+    foldr (_+_ ∘ (a ∙_) ∘ (uncurry _*ₗ_) ∘ < f , id >) 0# basisSet
+      ≈⟨ foldr-cong (λ y≈z _ → +-cong ∙-comm-*ₗ y≈z)
+                    (reflexive Eq.refl) basisSet ⟩
+    foldr (_+_ ∘ (uncurry _*_) ∘ < f , (a ∙_) >) 0# basisSet
+      ≈⟨ foldr-cong (λ y≈z x → +-cong (*-comm (f x) (a ∙ x)) y≈z)
+                    (reflexive Eq.refl) basisSet ⟩
+    foldr (_+_ ∘ (uncurry _*_) ∘ < (a ∙_) , f >) 0# basisSet
+      ≈⟨ sym (foldr-homo (a ∙_) basisSet) ⟩
+    f (foldr (_+ᴹ_ ∘ (uncurry _*ₗ_) ∘ < (a ∙_) , id >) 0ᴹ basisSet)
+      ≈⟨ ⟦⟧-cong (Setoid.sym ≈ᴹ-setoid (basisComplete)) ⟩
+    f a ∎)
+    where
+    v′ = lmToVec lm
+
+  -- Inner product extensional equivalence.
+  x·z≈y·z⇒x≈y : ∀ {x y} → DoubleNegationElimination ℓm →
+                 ∃₂ (λ s z → ((s *ₗ (x +ᴹ -ᴹ y) ≈ᴹ z) × (f z ≉ 0#))) →
```

Comment:
Any reason for declaring this as a dependent pair and not just a standard quantification?

Reply:
The name of this property mentions `z` (so one thinks it's an input), but the output doesn't have the right 'shape' as expected from the name.

I would have expected the type to be
```agda
∀ {x y} → (∀ {z} → x · z ≈ y · z) → x ≈ y
```